File: reactive_mongodb_example_java/src/main/java/com/github/nkolytschew/data/document/RatingDocument.java
package com.github.nkolytschew.data.document;

import org.hibernate.validator.constraints.NotEmpty;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.util.Date;
import java.util.List;
import java.util.UUID;

import javax.validation.constraints.NotNull;

@Document(collection = "rating")
public class RatingDocument {

    @Id
    private String id = UUID.randomUUID().toString();

    @Field
    @NotEmpty
    private String configurationId;

    @Field
    @NotNull
    private List<RatingComment> ratingCommentList;

    @Field
    @NotNull
    private Date creationDate = new Date();

    @Field
    private Boolean active = true;

    // jpa
    public RatingDocument() {
    }

    public RatingDocument(String configurationId, List<RatingComment> ratingCommentList) {
        this.configurationId = configurationId;
        this.ratingCommentList = ratingCommentList;
    }

    public RatingDocument(String id, String configurationId, List<RatingComment> ratingCommentList, Date creationDate, Boolean active) {
        this.id = id;
        this.configurationId = configurationId;
        this.ratingCommentList = ratingCommentList;
        this.creationDate = creationDate;
        this.active = active;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getConfigurationId() {
        return configurationId;
    }

    public void setConfigurationId(String configurationId) {
        this.configurationId = configurationId;
    }

    public List<RatingComment> getRatingCommentList() {
        return ratingCommentList;
    }

    public void setRatingCommentList(List<RatingComment> ratingCommentList) {
        this.ratingCommentList = ratingCommentList;
    }

    public Date getCreationDate() {
        return creationDate;
    }

    public void setCreationDate(Date creationDate) {
        this.creationDate = creationDate;
    }

    public Boolean getActive() {
        return active;
    }

    public void setActive(Boolean active) {
        this.active = active;
    }
}

File: reactive_mongodb_example/src/main/kotlin/com/github/nkolytschew/data/document/RatingDocument.kt
package com.github.nkolytschew.data.document

import org.hibernate.validator.constraints.NotEmpty
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field
import java.util.*
import javax.validation.constraints.Max
import javax.validation.constraints.Min
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

@Document(collection = "rating")
class RatingDocument(
    @Id val id: String? = UUID.randomUUID().toString(),
    // for every configuration, there should only be one rating-document
    @Field @NotEmpty val configurationId: String,
    @Field @NotNull val ratingCommentList: MutableList<RatingComment>,
    @Field var active: Boolean? = true,
    @Field var creationDate: Date? = Date()
)

class RatingComment() {
  @Id var id: String? = UUID.randomUUID().toString()
  @Field @NotEmpty lateinit var email: String
  @Field @NotEmpty lateinit var author: String
  @Field @NotNull var creationDate: Date? = Date()

  @Field @Min(1) @Max(10) @NotEmpty var rating: Int = 0
  @Field @NotEmpty @Size(min = 10) lateinit var description: String
  // set default to true
  @Field var active: Boolean? = true

  // constructors are for convenience
  constructor(email: String,
              author: String,
              rating: Int,
              description: String) : this() {
    this.email = email
    this.author = author
    this.rating = rating
    this.description = description
  }

  constructor(id: String,
              email: String,
              author: String,
              rating: Int,
              description: String) : this(email, author, rating, description) {
    this.id = id
  }


}